### DISCUSSION TITLE
Raw scancode support for key events (controller support)

### DISCUSSION THREAD

**Discussion by Paril** [mikke89/RmlUi#641] [Ideas]
I wanted to ask about this first before opening any bug report about this.

We're implementing menus for consoles, and realizing that one missing feature at the moment is that we have no obvious way of processing buttons within Rml's own events that don't map to an existing KeyIdentifier. For instance, we have Cancel and Confirm hooked up to Escape and Return respectively, but on a controller, the Triangle/Y Button doesn't really map to any KI, and we'd like to be able to receive key events for them.

I'm seeing two approaches here:
a) we create a custom event (like `custom-joydown` / `custom-joyup`) and pass that through.
b) we pick a range of unused KeyIdentifier values and use them for our own purposes; the enum doesn't have a Total/Count value so I don't think it's a closed range, but I worry about future updates stomping our range

What do you think would be the best approach for this?

**Comment by mikke89** [comment]
Right, good question. I don't have a strong preference here, I would say whatever works for you, but using a custom key identifier value looks easier at first glance.

I understand the concern about the range. Although, we haven't really added any new keys here at all, so I wouldn't be too worried about that. But if it would ease your mind, we could easily add a `KI_FIRST_CUSTOM_KEY` (naming suggestion very much welcome!). That seems very reasonable, same thing we do for property IDs.

**Reply by Paril** [reply]
That would be suitable for me, yeah. Just a guarantee that we can pick a range. For something like analog movement we'd have to use an event probably, but buttons should be good through the key API.

**Reply by mikke89** [reply]
Just implemented this solution here: https://github.com/mikke89/RmlUi/commit/5f2a0b022dd24a325e9cef058aba8affe3371c34
